language/move-stdlib/sources/struct_tag.move

```diff
@@ -0,0 +1,55 @@
+/// Module to decompose a move struct into it's components.
+module std::struct_tag {
+    use std::ascii::{Self, String};
+
+    struct StructTag has copy, store, drop {
+        /// Address of the entity that the struct belongs to.
+        /// taking `00000000000000000000000000000001::option::Option<u64>` for example,
+        /// the address will be `00000000000000000000000000000001`
+        address_: address,
+        /// the name of the module where the struct is defined.
+        /// using the example struct above the module name should be `option`
+        module_name: String,
+        /// the name of the struct itself.
+        /// using the example struct above the module name should be `Option`
+        struct_name: String,
+        /// the generics or tyepe params of the struct.
+        /// using the example struct above the module name should be `vector["u64"]`
+        generics: vector<String>
+    }
+
+    /// Returns the tag of the struct of type `T`
+    public native fun get<T>(): StructTag;
+
+    // Converts `self` into a tuple of it's inner values
+    public fun into(self: &StructTag): (address, String, String, vector<String>) {
+        (self.address_, self.module_name, self.struct_name, self.generics)
+    }
+
+    /// Returns the module authority for the struct of type `T`
+    public fun module_authority<T>(): StructTag {
```

Comment:
How to understand the module authority? I see the struct_name is hardcode `Witness`, so is it a programming pattern?

Reply:
Yeah exactly! My intention is that every module should be able to assert their ownership in this way. For example you could do something like:

```
public fun do_something<T: drop>(witness: T, object: &Object) {
   assert!(struct_tag::module_authority<Object> == struct_tag::get<T>, ENO_MODULE_AUTHORITY);
...
```

This is stating that, in order for this function to continue, we need a witness, and that Witness must be the module-authority witness produced by the same module that produced the Object.

This way we can guarantee that the function-call originated from the same module that defined Object in the first place.